src/main/java/com/visenze/datatracking/DataTracking.java

```diff
@@ -0,0 +1,164 @@
+package com.visenze.datatracking;
+
+
+
+import android.content.Context;
+import android.util.Log;
+
+import com.visenze.datatracking.data.DataTrackingResponse;
+import com.visenze.datatracking.data.DeviceData;
+import com.visenze.datatracking.data.EventData;
+import com.visenze.datatracking.data.EventsBody;
+import com.visenze.datatracking.data.ResultData;
+import com.visenze.datatracking.utils.Utils;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class DataTracking {
+
+    public static final String TAG = "DataTracking";
+    public static final long SESSION_TIMEOUT = 900000; //15 min
+
+    private static DataTracking mInstance;
+
+    private APIService apiService;
+    private String code;
+    private String uid;
+    private long sid;
+
+    private List<EventData> events;
+    private DataCollection dataCollection;
+    private DeviceData mDeviceData;
+
+    private DataTracking() {
+        apiService = HttpInstance.getRetrofitInstance(Constants.BASE_URL).create(APIService.class);
+        events = new ArrayList<EventData>();
+    }
+
+
+    public static synchronized DataTracking getInstance() {
+        if(mInstance == null) {
+            mInstance = new DataTracking();
+
+        }
+        return mInstance;
+    }
+
+
+    public void init(Context context, String code) {
+        String uid = getUid(context);
+        this.code = code;
+        initDataTracking(context, uid);
+    }
+
+    public void init(Context context, String code, String uid) {
+        if(uid == null || uid.isEmpty()) {
+            uid = getUid(context);
+        }
+        this.code = code;
+        initDataTracking(context, uid);
+    }
+
+
+    private String getUid(Context context) {
+        String uid = Utils.getUid(context);
+        if(uid == null || uid.isEmpty()) {
+            uid = UUID.randomUUID().toString();
```

Comment:
abstract the method into generate UID.

That getUid will call generateUid internally.
Why the method is "private" ? Should it be public ?

Reply:
also for getting Uid, do we still need the context? For first time init, we can extract the Application Context and use the sharedpreferences ?

Reply:
i think we just the reference of the SharedPreferences ? subsequent calls, we should not need the context anymore. Like this: https://github.com/visenze/visearch-sdk-android/blob/0424fa7a8f7bb8dd99f512697c151f136b972752/visearch-android/src/main/java/com/visenze/visearch/android/util/ViSearchUIDManager.java

Reply:
keep it as private because. generateUid will never be used by any other method. 
in my understanding keep a reference of context in a long lasting class in general is a very bad idea. in almost all case, context or context related object should try the best not be stored as a variable in an object. 
becuase activity context have all ui objects and is large in memory size. it will prevent the activity been garbage collected.

Reply:
yes I mean don't keep the context, only keep the reference to SharedPreferences. The getUid() method is useful to have and can be public. The generateUid should be kept private.

Reply:
added sessionManager